### DISCUSSION TITLE
MIDI to OSC | Shapekeys (Blendshapes)

### DISCUSSION THREAD

**Discussion by marcus-universe** [vrchat-community/osc#114] [Q&A]
Hi everyone 😉✌
  
I'm working on my own VRChat Piano Avatar. 
I created for every note on the piano a shape key (Blendshape) that presses the note down.
 
![image](https://user-images.githubusercontent.com/26997041/159889203-3d8719db-4bb8-4715-a79d-c95c21772eeb.png)
![image](https://user-images.githubusercontent.com/26997041/159889091-bc6537a6-adc5-45f5-a71c-7b8b34731d78.png)

I also received over Touch Designer Midi signals to OSC on Port 9000. The Signals provide a Midi Name (ADDRESS) that starts with /ch1n and after that with the corresponding number for the Midi Key.  Every signal provides a Float value between 0 - 1 (depending on how hard I press a key on the piano). 

![image](https://user-images.githubusercontent.com/26997041/159895744-b8f5bfcf-cf08-4eb4-b823-b10c024eddaa.png)
![image](https://user-images.githubusercontent.com/26997041/159899657-a8905377-0e7c-4cb4-8a2c-61dde5b4c953.png)

### Here a Spreadsheet for orientation 😉👍
![image](https://user-images.githubusercontent.com/26997041/159894970-505dec95-e479-4f22-9d34-e11b6d3d3a70.png)

In Unity, I created an Avatar Parameter Expression to Test the Key ch1n73 (C#) 😏

![image](https://user-images.githubusercontent.com/26997041/159897079-5d641cc2-c8dd-4f2f-a9d1-e44a63047aa7.png)

![image](https://user-images.githubusercontent.com/26997041/159896961-c329c746-5a03-434c-8c41-b37573693464.png)

**Now my question is, How I can change a shape key with the parameter values?** 🤔
Blendshapes have values between 0 to 100. This means I need to multiply it by 100. I found a doc for manipulating Shapekeys with C#.
[https://docs.unity3d.com/Manual/BlendShapes.html](https://docs.unity3d.com/Manual/BlendShapes.html)
**But how do I get the values from the OSC parameters?** 🤔

**Comment by EIA485** [comment]
if i were you id remap each note to a bool, this can easily be done in td. while this would not take velocity into effect implementing that would be fairly complicated. in order to turn those bools into the shapekeys moving you want to use the fx layer animator on your avatar. for each key make a down animation then have a layer in the animator go between entry, down and exit with the condition being that note's bool. if you need a more detailed explanation lmk.

**Reply by marcus-universe** [reply]
Thanks for the help 😊.
Then I have to figure out how to trigger Bools over OSC in TD ^^
I will try it tomorrow. It takes some time to make 88 Animations 😅✌
This is why I want to use the shape keys directly ^^

**Reply by EIA485** [reply]
in order to send bools from td use a math chop to clamp the midi vel to 0-1 instead of 0-127 then set the osc out chop to int.

**Reply by marcus-universe** [reply]
> in order to send bools from td use a math chop to clamp the midi vel to 0-1 instead of 0-127 then set the osc out chop to int.

Ok we have the Ints from 0-1. Now I making all the animation. :)
![image](https://user-images.githubusercontent.com/26997041/160100883-5e2b5eac-c01c-43b9-ab96-4b108b864d14.png)

**Reply by marcus-universe** [reply]
@EIA485 Where I have to add the condition and is it later possible to press multiple notes at the same time or do i need a blendtree? 😅
![image](https://user-images.githubusercontent.com/26997041/160105614-9384f791-a6bd-4962-a96a-7005314c03c1.png)

**Reply by EIA485** [reply]
you to it on the transition, im realizing you need to make a dummy default state for each layer. it is possible to press multiple notes, just make a layer for each one. remember to set all the layer's weights to 1.
also, remember to all all the notes as bool parameters both to the animater in the parameters tab and in the vrc parameters object.
[figure i should leave the unity docs here for reference](https://docs.unity3d.com/2019.4/Documentation/Manual/class-AnimatorController.html)

**Reply by marcus-universe** [reply]
@EIA485 **A small update:** I get it to work and it works fine. Thanks for the help :)

**For everyone else:** I figured out that VRChat has an [OscCore Package](https://github.com/vrchat/osccore/tree/all-in-one) that makes many things easier in C#. I guess this will be easier than making 88 Animation for On and Off. Instead of 176 Animation, it's possible to just route the shape keys to bools and multiply them by 10. [Here is the doc on how to manipulate shape keys in C#](https://docs.unity3d.com/Manual/BlendShapes.html)

**Reply by EIA485** [reply]
the only issue with this is that you cant have your own scripts on avatars, how ever you can use editor scripts to generate the animations and animators for you.
